File: Exercises2024/Chapter06_vle_free_energy/C/1-Gibbs/Source/eneri.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <math.h>
#include "system.h"
#include "conf.h"   

int Eneri(double Xi, double Yi, double Zi, int I, int Jb, double *En, double *Vir, int BoxID)
{
  double Dx, Dy, Dz, R2, Virij, Enij;
  int  J;
  
  *En = 0.0;
  *Vir = 0.0;
  for(J=Jb;J<Npart;J++)
    {
      if(Id[J]==BoxID)
	{
	  if(J!=I)
	    {
	      Dx = Xi - X[J];
	      Dy = Yi - Y[J];
	      Dz = Zi - Z[J]; 
	      if(Dx>Hbox[BoxID])
		{
		  Dx = Dx - Box[BoxID];
		}
	      else if(Dx<-Hbox[BoxID])
		{
		  Dx = Dx + Box[BoxID];
		}
	      if(Dy>Hbox[BoxID])
		{
		  Dy = Dy - Box[BoxID];
		}
	      else if(Dy<-Hbox[BoxID])
		{
		  Dy = Dy + Box[BoxID];
		}
	      if(Dz>Hbox[BoxID])
		{
		  Dz = Dz - Box[BoxID];
		}
	      else if(Dz<-Hbox[BoxID]) 
		{
		  Dz = Dz + Box[BoxID];
		}
	      R2 = Dx*Dx + Dy*Dy + Dz*Dz;
	      Ener(&Enij, &Virij, R2, BoxID);
	      *En+= Enij;
	      *Vir+= Virij;
	    }
	}
    }
  
  return 0;
}

File: Exercises2024/Chapter06_vle_free_energy/Fortran/1-Gibbs/Source/eneri.f
      Subroutine Eneri(Xi, Yi, Zi, I, Jb, En, Vir, BoxID)
      Implicit None

      Include 'parameter.inc'
      Include 'conf.inc'
      Include 'system.inc'
 
      Double Precision Xi, Yi, Zi, En, Dx, Dy, Dz, R2, Vir, Virij, Enij
      Integer I, J, Jb, BoxID

      En = 0.D0
      Vir = 0.D0
      Do J = Jb, Npart
         If (Id(J).Eq.BoxID) Then
            If (J.Ne.I) Then
               Dx = Xi - X(J)
               Dy = Yi - Y(J)
               Dz = Zi - Z(J)
               If (Dx.Gt.Hbox(BoxID)) Then
                  Dx = Dx - Box(BoxID)
               Else
                  If (Dx.Lt.-Hbox(BoxID)) Dx = Dx + Box(BoxID)
               End If
               If (Dy.Gt.Hbox(BoxID)) Then
                  Dy = Dy - Box(BoxID)
               Else
                  If (Dy.Lt.-Hbox(BoxID)) Dy = Dy + Box(BoxID)
               End If
               If (Dz.Gt.Hbox(BoxID)) Then
                  Dz = Dz - Box(BoxID)
               Else
                  If (Dz.Lt.-Hbox(BoxID)) Dz = Dz + Box(BoxID)
               End If
               R2 = Dx*Dx + Dy*Dy + Dz*Dz
               Call Ener(Enij, Virij, R2, BoxID)
               En = En + Enij
               Vir = Vir + Virij
            End If
         End If
      End Do

      Return
      End
